Merging client SDK work

angrbrd opened the pull request:
## Description

PR containing the client SDK work.
## Motivation and Context

This pull request contains all of the client sdk work, including all updated unit tests, Makefile updates, and documentation updates.
## How Has This Been Tested?

This pull request introduces or updates three unit test files: registrar.js, chain-test.js, asset-mgmt.js. chain-test.js and asset-mgmt.js may be run in development or network mode for deployment. All tests were automated with the "make node-sdk-unit-tests" command.
## Checklist:
- [x] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff)
- [x] Either no new documentation is required by this change, OR I added new documentation
- [x] Either no new tests are required by this change, OR I added new tests
- [x] I have run [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports), [go vet](https://golang.org/cmd/vet/), and [golint](https://github.com/golang/lint). I have cleaned up all valid errors and warnings in code I have added or modified. These tools may generate false positives. Don't be worried about ignoring some errors or warnings. The goal is clean, consistent, and readable code.

Signed-off-by: Anna Derbakova adderbak@us.ibm.com

Commit:
Add chaincode deploy UT

Adding a chaincode deploy UT in network mode. Commenting all the
following tests out for the time being (with test.skip) to work out
remaining issues with the deploy functionality.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Add basic chaincode deploy transaction path

Adding in the basic chaincode deploy transaction path, with
DeployTransactionCallback. An empty transaction structure is created,
sent over to the peer, and a response is returned.

Singed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Adding sdk_util.ts

Adding sdk_util.ts, which contains the hash computation functions as
well as the GenerateUUID() and GenerateTimestamp() functions. Updating
the ts compilation file tsconfig.json appropriately.

Signed-off-as: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Updating .gitignore file

Updating .gitignore file so we do not track the resulting
sdk_util.d.ts, sdk_util.js, and sdk_util.js.map files. We had two
.gitignore files, consolidating into one.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Add complete chaincode deploy transaction path

Adding the full chaincode deploy transaction path. The transaction is
sent to the peer, the deploy operation succeeds, and a txUUID is
returned.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Deflate the size of the deployment payload

Implement sdk_util.GenerateTarGz function which filters out unnecessary
file types from the deployment transaction, significantly reducing the
size of the payload.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Propagate chaincode hash up

Propagate the chaincode hash value up to the calling UT code. That way,
this value can be stored and used in the subsequent invoke and query
tests.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Updating UTs

Updating UTs to include a sleep after the chaincode deployment.
Otherwise, subsequent tests fail as the chaincode Docker container does
not have enough time to come up. All UTs are now passing.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Move to using sdk_util methods for UUID and Timestamp

Move the UUID and Timestamp generation methods to sdk_util.ts rather
than having them all lumped into hlc.ts.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Removing console.log() debug statements

Replacing all instances of console.log() with debug().

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Updating README.md

Updating the README.md to include new instructions to deploy chaincode
in network mode.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
FIxing error with process.env call

tsc was failing on process.env.GOPATH. Looks like it expects
process.env[‘GOPATH’] instead.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Updating .gitignore and package.json

Updating .gitignore to include the resulting .js build files.
Updating package.json to include “tar-fs” module.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Updating UTs

Updating the chain-tests.js UTs to have a standardized
testChaincodePath location, so that the user running the tests will not
have to update the UTs. They will just have to place their project
files into a specific location. The README.md will be updated to match
the new location.

Updating the asset-mgmt.js test case to have it deploy properly in
network mode. The deploy operation succeeds, but the last test still
fails due to the chaincode hash being passed up incorrectly. Therefore,
the query operation gets the wrong hash and fails. That will be fixed
shortly.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Updating README.md

Updating the README.md file to include more detailed instructions on
how to run the UTs.

Singed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Merging in Keith's changes

Adding Keith’s changes to hlc.ts and UTs.

Making the necessary modifications to make all tests functional for
deployment in network mode, NOT in development mode. This requires
properly setting up the chaincode projects according to instructions.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Updating UT instructions for deploy in network mode

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Adding deployment in development mode + UTs

Adding implementation for chaincode deployment in development mode.
Updating chain-tests.js unit test to run in both development and
network mode based on the $DEPLOY_MODE environment variable. Updating
README.md accordingly.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Latest changes to hlc.ts and UT files

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Adding run-unit-tests.sh

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Current status of UT asset-mgmt-with-roles.js

SIgned-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Including Makefile and updated run-unit-tests.sh

Updating UT files to work together with run-unit-tests.sh.

Updating README.md for most up to date instructions.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Adding auto-generated documentation and compiled .js files. Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

srderson commented on sdk/node/doc/classes/_crypto_.crypto.html:
```diff
@@ -157,7 +157,7 @@
 							<aside class="tsd-sources">
 								<ul>
-									<li>Defined in <a href="https://github.com/smithbk/fabric/blob/2d4b64c/sdk/node/src/crypto.ts#L78">crypto.ts:78</a></li>
+									<li>Defined in <a href="https://github.com/angrbrd/fabric/blob/7e4cfbe/sdk/node/src/crypto.ts#L78">crypto.ts:78</a></li>
```
Why are there lots of links to your personal repository in these doc files?

angrbrd commented:
That's a good question... These are auto-generated files and I'm not 100% sure how that ends up in there. But I see your point. I'll look into into this and see what can be done.

binhn commented on sdk/node/doc/classes/_hlc_.eventdeploycomplete.html:
```diff
@@ -0,0 +1,282 @@
+						<li class="state loading">Preparing search index...</li>
+						<li class="state failure">The search index is not available</li>
+					</ul>
+					<a href="../index.html" class="title">Hyperledger OpenBlockChain</a>
```
we shouldn't use this name `Hyperledger OpenBlockChain`

binhn commented on sdk/node/doc/classes/_hlc_.eventdeploysubmitted.html:
```diff
@@ -0,0 +1,265 @@
+	<meta charset="utf-8">
+	<meta http-equiv="X-UA-Compatible" content="IE=edge">
+	<title>EventDeploySubmitted | Hyperledger OpenBlockChain</title>
+	<meta name="description" content="">
```
Many instances of `Hyperledger OpenBlockChain`. 
Change it to `Hyperledger Fabric`

adecaro commented:
Hi @angrbrd, how did you test deploy in network mode? In particular, how did you setup the fabric network? Actually, I'm having trouble in setting up the fabric using vagrant. Nodejs, npm and typings versions are old.

BTW, I have seen that there are some dependencies missing in the file package.json like protobujs and lodash. Am I missing something?

smithbk commented:
Anya, to fix the "Hyperledger OpenBlockChain. " problems and to make the generated doc less verbose, replace the typedoc command in fabric/sdk/node/makedoc.sh with the following:

typedoc -m amd \
    --name 'Node.js Hyperledger Fabric SDK' \
        --includeDeclarations \
        --excludeExternals \
        --excludeNotExported \
    --out doc \
    src/hlc.ts typedoc-special.d.ts

Run makedoc.sh
git add makedoc.sh doc
git commit & push

smithbk commented:
@adecaro The sdk/node/bin/run-unit-tests.sh script runs tests in both network mode and dev mode.  You can look at that script to see how to set up the directory structure for deploy in network mode.  See the prepareExampleForDeployInNetworkMode function.  Maybe we should provide that as a standalone script to make it easier for users to set up other examples for deploy.

smithbk commented:
@adecaro I should have mentioned that setting up the directory structure is described in the README.md

Commit:
Updating auto-generated documentation and makedoc.sh

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

adecaro commented:
@smithbk Sorry, I'm confused. _run-unit-tests.sh_ is not mentioned in the README file. 

I got how to prepare the folder structure to deploy a chaincode in network mode. That's fine.

Anyway, even before that, when I run _make node-sdk_ in a clean vagrant instance, it just fails with the following message:

```
vagrant@ubuntu-1404:/opt/gopath/src/github.com/hyperledger/fabric$ make node-sdk
cp ./protos/*.proto ./sdk/node/lib/protos
cp ./membersrvc/protos/*.proto ./sdk/node/lib/protos
cd ./sdk/node && npm install && sudo npm install -g typescript && sudo npm install typings --global && typings install
npm WARN package.json hlc@0.0.9 No repository field.
npm WARN package.json hlc@0.0.9 No license field.
npm ERR! Linux 3.13.0-86-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
npm ERR! node v0.12.7
npm ERR! npm  v2.11.3
npm ERR! path ../har-validator/bin/har-validator
npm ERR! code EPROTO
npm ERR! errno -71

npm ERR! EPROTO, symlink '../har-validator/bin/har-validator'
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>
npm ERR! Linux 3.13.0-86-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
npm ERR! node v0.12.7
npm ERR! npm  v2.11.3
npm ERR! path npm-debug.log.b14eda52099085229a86ec264b001337
npm ERR! code ETXTBSY
npm ERR! errno -26

npm ERR! ETXTBSY, rename 'npm-debug.log.b14eda52099085229a86ec264b001337'
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /opt/gopath/src/github.com/hyperledger/fabric/sdk/node/npm-debug.log
make: *** [node-sdk] Error 185
```

Any idea on what I'm missing?

angrbrd commented:
@adecaro the run-unit-tests.sh script is what gets triggered by the `make node-sdk-unit-tests` command. That command is described in the README.md. If you take a look at that run-unit-tests.sh script, it will run through chaincode deployments in both network mode as well as development mode.

I am setting up a clean vagrant environment right now based on my branch with these changes. I will touch base shortly.

adecaro commented:
Thanks @angrbrd :)

rameshthoomu commented:
@angrbrd: Travis builds are failing with npm command not found error message as the npm is not installed in travis. Updated below npm command and it worked: 
`cd ./sdk/node && sudo apt-get install -y npm && sudo npm install -g typescript && sudo npm install typings --global && typings install`. 

I could see an issue with membersrvc failed to start: Below is the error message:
`Documentation generated at /home/travis/gopath/src/github.com/hyperledger/fabric/sdk/node/doc
ERROR: member services failed to start; see /tmp/node-sdk-unit-test/membersrvc.log`
**Travis Build for reference:**
"https://travis-ci.org/rameshthoomu/fabric/builds/137870434"

rameshthoomu commented:
Regarding this issue: membersrvc image is not created before run node-sdk-tests. We have create membersrvc image first and then run node-sdk-tests. Reference Makefile: `unit-test: peer-image gotools node-sdk-unit-tests`. We have to update above command to run `make membersrvc` and `peer-membersrvc-image.`

Commit:
Fixing the usage of request.chaincodeName

Verifying that request.chaincodeName is used correctly for deploy in
development mode, as opposed to request.chaincodeID.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

angrbrd commented:
@adecaro I just pushed another change should hopefully address any issues you are seeing. The things you need to do in a clean vagrant environment to run the unit tests from this branch are below. Please give this a try and let me know if it works for you.

```
cd $GOPATH/src/github.com/hyperledger/fabric/membersrvc
go build
cd $GOPATH/src/github.com/hyperledger/fabric/peer
go build
cd $GOPATH/src/github.com/hyperledger/fabric
make node-sdk-unit-tests
```

Commit:
Updating Makefile to address Travis CI failures

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

Commit:
Adding install of npm to Makefile

npm needs to be installed for the Node.js unit tests to pass.

Signed-off-by: Anna D Derbakova <adderbak@us.ibm.com>

smithbk commented:
CI is still failing here?  Is anyone investigating?

rameshthoomu commented:
@angrbrd is looking these issues.

angrbrd commented:
@binhn and I discussed and agreed that we will merge this without Travis running at this time. The issue with Travis is not obvious and all the unit tests are passing locally. I have created another branch that contains identical changes, but removes the Travis CI integration. Therefore, I will now close this PR and open a new one from that new branch.
